Let simulation start server with custom ClientManager

negedng opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md

Does the changelog need to be updated?
See: https://github.com/adap/flower/blob/main/doc/source/changelog.rst
-->

#### Reference Issues/PRs
Can't start simulation with custom ClientManager.
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create a link to the issues or pull requests
you resolved.
-->

#### What does this implement/fix? Explain your changes.
Adding extra parameters to default server initialization and simulation start to let users add a custom ClientManager instead of always using the SimpleClientManager.
<!--
Explain why this PR is needed and what kind of changes have you done.
Example: the variable rnd was not clear and therefore renamed to fl_round. 
-->

#### Any other comments?


<!--
Please be aware that it may take some time until we can check this PR. 
If you have an urgent request or question please use the Flower Slack channel.
The Slack channel is really active and contributors respond pretty fast. 

We value your contribution and are aware of the time you put into this PR.
Therefore, thank you for your contribution. 
-->

tanertopal commented:
@negedng thank you for the PR! Could you additionally prepare a changelog entry?

@danieljanes this looks good to me. I don't see any downsides of passing a custom client manager. Wdyt?

negedng commented:
@tanertopal I'm not sure what do you mean by adding a changelog entry. I updated the `doc/source/changelog.rst`, I hope it is the right way.

Also, one of the tests showed a formatting issue that I changed. I hope it is better now!

danieljanes commented on src/py/flwr/simulation/app.py:
```diff
@@ -96,6 +100,10 @@ def start_simulation(  # pylint: disable=too-many-arguments
         `flwr.server.strategy.FedAvg`.
+    client_manager: Optional[flwr.server.ClientManager] (default: None)
+        An implementation of the abstract base class `flwr.server.ClientManager`.
+        If no strategy is provided, then `start_server` will use
```
```suggestion
        If no implementation is provided, then `start_simulation` will use
```

danieljanes commented on src/py/flwr/simulation/app.py:
```diff
@@ -24,8 +24,10 @@
+from flwr.server.client_manager import ClientManager
 from flwr.server.history import History
 from flwr.server.strategy import Strategy
+
```
Should this empty line be removed?

danieljanes commented:
Thanks for the PR @negedng, looks good! I've just added two small comments.

@tanertopal I agree, let's merge this :)

tanertopal commented:
@negedng do you want to apply Daniels suggestions or shall I do? Afterwards we can merge it.

negedng commented:
Also, a test shows that there is a sorting issue in the imports for simulation/app.py, but I can't figure out the proper way to sort the imports. Can you have a look? thanks!

tanertopal commented on src/py/flwr/simulation/app.py:
```diff
@@ -149,7 +156,7 @@ def start_simulation(  # pylint: disable=too-many-arguments
     # Initialize server and server config
     config = {"num_rounds": num_rounds}
-    initialized_server, initialized_config = _init_defaults(None, config, strategy)
+    initialized_server, initialized_config = _init_defaults(None, config, strategy, client_manager)
```
```suggestion
    initialized_server, initialized_config = _init_defaults(
        None, config, strategy, client_manager
    )
```

negedng commented:
Hi, any update on this? Do I need to change something?

tanertopal commented:
@negedng There was an issue with PyPi at the time I wanted to merge it and CI got blocked iirc. I'll have a look today again and merge it if the CI goes through.

tanertopal commented:
@negedng there is still an issue in:

```src/py/flwr/server/app.py:97: error: Missing positional argument "client_manager" in call to "_init_defaults"```

could you have a look?

tanertopal commented on src/py/flwr/server/app.py:
```diff
@@ -55,6 +56,10 @@ def start_server(  # pylint: disable=too-many-arguments
             `flwr.server.strategy.FedAvg`.
+        client_manager: Optional[flwr.server.ClientManager] (default: None)
+            An implementation of the abstract base class `flwr.server.ClientManager`.
+            If no implementation is provided, then `start_simulation` will use
```
```suggestion
            If no implementation is provided, then `start_server` will use
```

tanertopal commented on src/py/flwr/server/app.py:
```diff
@@ -94,7 +99,7 @@ def start_server(  # pylint: disable=too-many-arguments
     >>> )
     """
-    initialized_server, initialized_config = _init_defaults(server, config, strategy)
+    initialized_server, initialized_config = _init_defaults(server, config, strategy, client_manager)
```
```suggestion
    initialized_server, initialized_config = _init_defaults(
        server, config, strategy, client_manager
    )
```
